types/msgservice/validate.go

```diff
@@ -0,0 +1,116 @@
+package msgservice
+
+import (
+	"errors"
+	"fmt"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+
+	msg "cosmossdk.io/api/cosmos/msg/v1"
+)
+
+// ValidateAnnotations validates that the proto annotations are correct.
+// More specifically, it verifies:
+// - all services named "Msg" have `(cosmos.msg.v1.service) = true`,
+// - all sdk.Msgs have correct `(cosmos.msg.v1.signer) = "..."`.
+//
+// More validations can be added here in the future.
+//
+// If `protoFiles` is nil, then protoregistry.GlobalFile will be used.
+func ValidateProtoAnnotations(protoFiles *protoregistry.Files) error {
+	if protoFiles == nil {
+		protoFiles = protoregistry.GlobalFiles
+	}
+
+	var (
+		serviceErrs []error
+		messageErrs []error
+	)
+	protoFiles.RangeFiles(func(fd protoreflect.FileDescriptor) bool {
+		for i := 0; i < fd.Services().Len(); i++ {
+			sd := fd.Services().Get(i)
+			if sd.Name() == "Msg" {
+				// We use the heuristic that services name Msg are exactly the
+				// ones that need the proto annotations check.
+				err := validateMsgServiceAnnotations(protoFiles, sd)
+				if err != nil {
+					serviceErrs = append(serviceErrs, err)
+				}
+
+				for j := 0; j < sd.Methods().Len(); j++ {
+					err := validateMsgAnnotations(protoFiles, sd.Methods().Get(j).Input())
+					if err != nil {
+						messageErrs = append(messageErrs, err)
+					}
+				}
+			}
+		}
+
+		return true
+	})
+
+	return errors.Join(errors.Join(serviceErrs...), errors.Join(messageErrs...))
+}
+
+// validateMsgServiceAnnotations validates that the service has the
+// `(cosmos.msg.v1.service) = true` proto annotation.
+func validateMsgServiceAnnotations(protoFiles *protoregistry.Files, sd protoreflect.ServiceDescriptor) error {
+	ext := proto.GetExtension(sd.Options(), msg.E_Service)
+	isService, ok := ext.(bool)
+	if !ok {
+		return fmt.Errorf("expected bool, got %T", ext)
+	}
+
+	if !isService {
+		return fmt.Errorf("service %s does not have cosmos.msg.v1.service proto annotation", sd.FullName())
+	}
+
+	return nil
+}
+
+// validateMsgAnnotations validates the sdk.Msg service has the correct
+// `(cosmos.msg.v1.signer) = "..."` proto annotation.
+func validateMsgAnnotations(protoFiles *protoregistry.Files, md protoreflect.MessageDescriptor) error {
+	ext := proto.GetExtension(md.Options(), msg.E_Signer)
```

Comment:
Can we just use the x/tx `GetSignersContext` to do this validation? Seems like there's a lot of repeated logic

Reply:
Where is GetSignersContext initialized in runtime? If we do that way, let's pass that same instance of GetSignersContext to this validation, so that it can fully pre-populate the cache and be faster in antehandlers.

Pushing this idea further, we can also remove this msg validation here, and just do it in `NewGetSignersContext` which does validation+population in one go.